conformance/README.md

```diff
@@ -172,66 +172,106 @@ ensuring that all input forms handle scenarios the same way.
+> the encoder by providing AST fragments and expecting certain bytes.
+> I can imagine that gets tricky given the variety of encoding options available.
+> Perhaps we could have DSL clauses that direct specific encoding choices, so that
+> we can expect specific byte sequences.
 
-* A test runner can conceivably transcribe every test document into both text
-  and binary.
-* The test suite would inherently exercise both parsers *and both encoders*.
 
-When the implementation does not work with ASTs, those fragments can be
-near-trivially be transcoded into text fragments.
+## Ion versions
 
-It would also be valuable if the DSL can be extended to intentionally focus on
-the encoder by providing AST fragments and expecting certain bytes.
-I can imagine that gets tricky given the variety of encoding options available.
-Perhaps we could have DSL clauses that direct specific encoding choices, so that
-we can expect specific byte sequences.
+The examples above illustrate the `ion_1_0` and `ion_1_1` entry points.
+These are derived forms, shorthand for the common extensions to a more primitive
+starting point: an empty document.  Here's how that is expressed:
 
+```
+(document (produces))
+```
 
-## Ion versions
+The `document` clause is the true root of all tests.
+It starts a test case with an empty document, no IVM, no bytes of any kind.
+The body of the clause is either an expectation or some extensions with more
+data.
+The example above simply says that an empty Ion document produces no data.
 
-The examples above illustrate the `ion_1_0` and `ion_1_1` entry points.
-In addition, the `ion_1_x` form declares behavior common to _both_ 1.0 and 1.1.
+To make a more meaningful test, we must add some input to the document:
 
 ```
-(ion_1_x (text "1::true")
-         (signals "Invalid annotation"))
+(document (text "null.int") (denotes (null int)))
 ```
 
-To be more specific, `(ion_1_0 _form_ ...)` behaves like:
+Here, the input document is an eight-byte text document containing exactly the
+given characters, and the test expects the implementation to produce a null `int`.
+
+The `ion_1_*` clauses are shorthands for extending the empty document with Ion 
+version markers.
+To be more specific, `(ion_1_0 _form_ ...)` is equivalent to:
 
 ```
-(each (text)              // Text input with no IVM
-      (text "$ion_1_0")
-      (binary "E00100EA")
-      (then _form_ ...))
+(document
+  (then (ivm 1 0) _form_ ...))
 ```
 
 `(ion_1_1 _form_ ...)` is equivalent to:
 
 ```
-(each (text "$ion_1_1")
-      (binary "E00101EA")
-      (then _form_ ...))
+(document
+  (then (ivm 1 1) _form_ ...))
 ```
 
-`(ion_1_x _form_ ...)` is equivalent to:
+Finally, `(ion_1_x _form_ ...)` is equivalent to:
+
 ```
-(each (text)
-      (text "$ion_1_0")
-      (binary "E00100EA")
-      (text "$ion_1_1")
-      (binary "E00101EA")
-      (then _form_ ...))
+(document
+  (then (ivm 1 0) _form_ ...)
+  (then (ivm 1 1) _form_ ...))
```

Comment:
Still equivalent to this, right?
```
(document (each (ivm 1 0) 
                (ivm 1 1) 
                (then _form_ )))
```

Reply:
Not quite: no fragment is required by `ion_1_x` but one _is_ required by `then`.  That would cause your proposed expansion to fail.

Why does `then` require a fragment? I want all extensions to be consistent whether they require fragments, and `each` makes no sense without at least one (else it would be effectively an empty-loop).  The need for no-fragment `then` seems limited, so I went for a consistent grammar at the cost of slightly more code for an edge case.

Reply:
Okay, so it's equivalent to one of these, depending on the remainder of the test case?
```
(document (each (ivm 1 0) 
                (ivm 1 1) 
                (then ... )))

(document (each (ivm 1 0) 
                (ivm 1 1) 
                (each ... )))


(document (each (ivm 1 0) 
                (ivm 1 1) 
                (_expectation_ ... )))
```

Reply:
I'm a little unclear on what you mean, but I think so? There's always multiple ways to write things.
 
The expansion in the text is something that works regardless of whether the `ion_1_*` body starts with a fragment.  I want to avoid a design that requires inspection of the body.

Reply:
Here's a context-free expansion of `(ion_1_x <body>)` using each:

```
(document (each (ivm 1 0) 
                (ivm 1 1) 
                (then (toplevel) <body>)))
```

Note the injection of an empty fragment to ensure the `then` clause always has one even if <body> is just an expectation.

Reply:
> I want to avoid a design that requires inspection of the body.

Okay, so I think there may be a fundamental difference between how we are approaching this. In your case, this DSL _is_ actually code, and you can do a (presumably) simple transformation. I am not doing any rewriting of the DSL because at the end of the day, this DSL is not a runnable thing in Java—I'm working with `IonReader` or `IonValue` and it needs to be converted or interpreted somehow—so I'm doing the expansion at the same time as I am converting it.

Reply:
That's not the case; my implementation is a straightforward CS200-level interpreter.  As the tests are being evaluated, it walks the S-expression tree recursively.  There's functions to recognize a "clause" (S-expression starting with a keyword), and a primitive dispatch mechanism using maps from keywords to evaluation functions. There's basically one dispatch map per nonterminal in the grammar, and an evaluation method per nonterminal-option.  There's no preprocessing of the clauses, everything happens on the fly.

The spec doesn't require any implementation to rewrite expressions, its simply defining the semantics of some features like `$ion_1_x` in terms of others in order to be clear and precise.

Side observation, now that I started implementing `denotes`: the more every grammar is a clause, the less code there is in the implementation.

Reply:
Great!  You came to the same conclusion I did regarding what's most sensible.